integration/tests/cook/util.py

```diff
@@ -43,46 +57,114 @@ class User(object):
+        self.previous_token = None
+
+    def __enter__(self):
+        global session
+        super().__enter__()
+        assert self.previous_token is None
+        self.previous_token = session.headers.get('Authorization')
+        session.headers['Authorization'] = self.auth_token
+
+    def __exit__(self, ex_type, ex_val, ex_trace):
+        global session
+        super().__exit__(ex_type, ex_val, ex_trace)
+        if self.previous_token is None:
+            del session.headers['Authorization']
+        else:
+            session.headers['Authorization'] = self.previous_token
+            self.previous_token = None
+
+
 class UserFactory(object):
     """
     Factory object used to create unique user names for a given test function.
     Usernames are composed of the test name and and increasing integer values.
     """
 
     def __init__(self, test_handle):
-        test_id = test_handle.id()
-        base_name = test_id[test_id.rindex('.test_')+6:].lower()
-        self.__user_generator = ( f'{base_name}_{i}' for i in range(1000000) )
+        # Select the authentication scheme
+        if http_basic_auth_enabled():
+            self.user_class = _BasicAuthUser
+        elif kerberos_enabled():
+            self.user_class = _KerberosUser
+        else:
+            raise NotImplementedError(f'Unsupported user authentication scheme: {_cook_auth_scheme()}')
+        # Set up generator for new user objects
+        if test_handle:
+            test_id = test_handle.id()
+            test_base_name = test_id[test_id.rindex('.test_')+6:].lower()
+            base_name = os.getenv('COOK_TEST_USER_PREFIX', f'{test_base_name}_')
+            self.__user_generator = ( f'{base_name}{i}' for i in range(1000000) )
 
     def new_user(self):
-        """Return a fresh User object."""
-        return User(next(self.__user_generator))
+        """Return a fresh user object."""
+        return self.user_class(next(self.__user_generator))
 
     def new_users(self, count=None):
-        """Return a sequence of `count` fresh User objects."""
-        return map(User, itertools.islice(self.__user_generator, 0, count))
+        """Return a sequence of `count` fresh user objects."""
+        return map(self.user_class, itertools.islice(self.__user_generator, 0, count))
 
+    @functools.lru_cache()
+    def default(self):
+        """Return the default user"""
+        return self.user_class(_get_default_user_name())
 
-# Default user for multi-user test runs
-default_user = User('root')
+    @functools.lru_cache()
+    def admin(self):
+        """Return the administrator user"""
+        name = os.getenv('COOK_ADMIN_USER_NAME', _default_admin_name)
+        return self.user_class(name)
+
+    @functools.lru_cache()
+    def impersonator(self):
+        """Return the impersonator user"""
+        name = os.getenv('COOK_IMPERSONATOR_USER_NAME', _default_impersonator_name)
+        return self.user_class(name)
```

Comment:
Can we check against `/settings` that the users chosen by these functions actually have the capabilities they're supposed to?

Reply:
That information isn't exposed in the `/settings` data, but the assumptions are documented in the README. There's always going to be some point where we just have to assume the person running the tests configured things correctly.

Reply:
Gotcha